modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala

```diff
@@ -1172,11 +1173,38 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
       s"Unhandled trait binding:\ntype: $tpe\nvalue: ${Node.printJson(node)}"
   }
 
-  private def unfoldTrait(tr: Trait): Hint.Native = {
+  private def unfoldTrait(shape: Shape, tr: Trait): Hint.Native = {
     val nodeAndType = NodeAndType(tr.toNode(), tr.toShapeId().tpe.get)
-    Hint.Native(recursion.ana(unfoldNodeAndType)(nodeAndType))
+    Hint.Native(
+      typedNode = recursion.ana(unfoldNodeAndType)(nodeAndType),
+      rawNode = nodeAndType.node,
+      tr = Type.Ref(tr.namespace, tr.name),
+      isRecursive = isRecursiveTraitOf(shape.getId, tr.toShapeId(), Set.empty)
+    )
   }
 
+  // returns true if there's a reference to `shape` in the trait closure of `id`.
+  private def isRecursiveTraitOf(
```

Comment:
This looks a bit like the recursive check in `DynamicModelCompiler`, which has known performance issues: #1258.

I'm using this implementation for now, as I don't expect trait trees to be as deep (or wide) as shapes used for data, and so far it seems like a sane assumption... that said, I'm happy to use a more optimized implementation if we can make one.

Notably, this isn't a simple `isRecursive(id)` - I'm checking `shape` in particular appears in the closure of the given trait.

Some examples:

- `documentation -> documentation` is true
- `documentation -> trait` is true (the opposite is also true, which makes me think this is commutative)
- `foo -> documentation` is false for any user-defined `foo`.

It doesn't affect any shapes in `bootstrapped`.

Here are some stdlib examples:

```scala
// Documentation
val hints: Hints = Hints(
  ShapeId("smithy.api", "documentation") -> Document.fromString("Adds documentation to a shape or member using CommonMark syntax."),
  ShapeId("smithy.api", "trait") -> Document.obj(),
)

// Trait
val hints: Hints = Hints(
  ShapeId("smithy.api", "documentation") -> Document.fromString("Makes a shape a trait."),
  ShapeId("smithy.api", "trait") -> Document.obj("selector" -> Document.fromString(":is(simpleType, list, map, structure, union)"), "breakingChanges" -> Document.array(Document.obj("change" -> Document.fromString("presence")), Document.obj("path" -> Document.fromString("/structurallyExclusive"), "change" -> Document.fromString("any")), Document.obj("path" -> Document.fromString("/conflicts"), "change" -> Document.fromString("update"), "severity" -> Document.fromString("NOTE"), "message" -> Document.fromString("Adding more conflicts to a trait could cause previously written models to fail validation.")))),
)

// HttpApiKeyAuth (references the above two and something else, for comparison)
val hints: Hints = Hints(
  smithy.api.Documentation("An HTTP-specific authentication scheme that sends an arbitrary API key in\na header or query string parameter."),
  smithy.api.Trait(selector = Some("service"), structurallyExclusive = None, conflicts = None, breakingChanges = Some(List(smithy.api.TraitDiffRule(change = smithy.api.TraitChangeType.REMOVE.widen, severity = smithy.api.TraitChangeSeverity.ERROR.widen, path = None, message = None)))),
  smithy.api.AuthDefinition(traits = None),
)
```

Reply:
> that said, I'm happy to use a more optimized implementation if we can make one.

I'd much rather we did that to be honest, and constructed eagerly a set of all shapes that match the recursion criteria.